🚀 4단계 - 자동차 경주(우승자)

giyeon95 opened the pull request:
안녕하세요 리뷰어님!

4단계 기능 구현 후 리뷰 요청 드립니다.
3단계 기능 구현시 리뷰 주신 내용도 이번 PR에서 함께 수정을 하였는데요,
혹시 잘못 이해한 부분이 있다면 말씀부탁드려요.

감사합니다.🙏

giyeon95 committed:
refactor: 불필요 TextFixture 제거 (리뷰 반영)

1. 넵 픽스쳐 부분에서 불필요하게 override가 발생하는 것 같아서 드린 코멘트였습니다.
2. 크게 두가지로 작성할 수 있을 것 같아요.

```kotlin
interface RacingCarRandom {
    fun nextInt(): Int
}

interface RacingCarEngine {
    fun canMove(): Boolean
}
```

어떤 방법이던 레이싱 카 도메인에 적합한 인터페이스를 도출해볼 수 있지 않을까요?

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1019852608_

giyeon95 committed:
test: 테스트코드 리뷰 반영

giyeon95 committed:
feat: 자동차 생성시 이름을 부여 받도록 기능 추가

giyeon95 committed:
feat: 우승자 추출 및 결과 반환 View 추가

ddaaac commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -1,23 +1,27 @@
 class Car(
-    val pos: Int = 0
+    val name: String,
+    val pos: Int = 0,
```
오 이제 ktlint도 trailing comma를 지원하나보군요!

ddaaac commented on src/main/kotlin/racingcar/RaceStage.kt:
```diff
@@ -1,22 +1,31 @@
-        fun create(carCount: Int): RaceStage = RaceStage(
-            List(carCount) { Car() }
+        fun create(carNames: List<String>): RaceStage = RaceStage(
+            List(carNames.size) { index -> Car(carNames[index]) }
```
kotlin은 [여러 collection의 trasforming](https://kotlinlang.org/docs/collection-transformations.html#map)을 지원합니다. 조금 더 적절한 방법이 있을 것 같아요!

ddaaac commented on src/main/kotlin/racingcar/RaceStage.kt:
```diff
@@ -1,22 +1,31 @@
     }
 
     fun racing(): RaceStage = RaceStage(
-        cars.map { car -> car.race(Random.Default) }
+        cars.map { car -> car.race(RacingCarEngine.canMove(((START_VALUE until LIMIT_VALUE).random()))) }
     )
```
이전 단계에서 [코멘트](https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1019852608)를 드렸던 가장 큰 이유는 도메인인 RaceStage#racing도 테스트가 가능해야하지 않을까 싶어서였어요. 그래서 인터페이스로 분리하는 건 어떨까하는 코멘트를 남겼던 거였구요.
이 함수를 테스트 가능하게 하려면 어떻게 구조를 잡아야할지 고민해보시면 좋을 것 같습니다 😄

giyeon95 replied:
개인적으로 코드를 작성하면서, RaceStage#racing 은 cars 를 순회하면서 Random 값을 생성하는 역할 외에는 다른 로직이 없다고 생각하였습니다, 그렇다보니 RaceStage#racing 에서도 테스트를 한다면, 중복된 테스트가 되는것은 아닌가? 싶은 고민이 들었었구요..ㅎㅎ!🤔
 > 이부분은 car#race 에서 테스트를 하고 있다고 생각 했었어요.😁

우선은 리뷰 주신 내용으로, CarStrategy를 이용하여, 테스트가 가능한 구조로 변경 후 추가 테스트 코드를 작성해보았는데요,
혹시 제가 잘못 이해하고 있는 내용이 있다면 말씀 주시면 좋을 것 같습니다.🙏

ddaaac replied:
지금은 코드가 간단하기도 하고, 직접 구현하셨으니 중복된 테스트라는 걸 알 수 있지만
코드가 복잡해지거나, 다른 사람이 봤을 때는 두 함수가 비슷한 역할을 하는지 모를 수도 있으니 작성하는 편이 더 좋지 않을까 싶었습니다.
교육과정에서 모든 도메인 함수에 대해서 테스트 작성을 권하는 것도 비슷한 맥락이구요 😄 

+) CarStrategy에서 바로 Engine을 리턴해도 되지 않을까요? 오히려 이 편이 조금 더 자연스러운 것 같기도 하구요.

ddaaac reviewed (changes requested):
안녕하세요 기연님 👋 

이전 단계에서 잘 구현해주셔서 코드에 큰 변화없이 이번 단계 요구사항을 달성하셨네요 👍 

몇 가지 코멘트를 남겼으니 확인 부탁드려요~

giyeon95 committed:
refactor: List 생성 방식 가독성 개선

kotlin은 [여러 collection의 trasforming](https://kotlinlang.org/docs/collection-transformations.html#map)을 지원합니다. 조금 더 적절한 방법이 있을 것 같아요!

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/844#discussion_r1020852138_

giyeon95 committed:
refactor: CarStrategy 를 이용한 RaceStage#racing 테스트 가능한 구조로 개선

이전 단계에서 [코멘트](https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1019852608)를 드렸던 가장 큰 이유는 도메인인 RaceStage#racing도 테스트가 가능해야하지 않을까 싶어서였어요. 그래서 인터페이스로 분리하는 건 어떨까하는 코멘트를 남겼던 거였구요.
이 함수를 테스트 가능하게 하려면 어떻게 구조를 잡아야할지 고민해보시면 좋을 것 같습니다 😄

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/844#discussion_r1020852891_

ddaaac commented on src/main/kotlin/racingcar/RaceStage.kt:
```diff
@@ -1,22 +1,29 @@
+        cars.map { car -> car.race(RacingCarEngine.canMove(strategy.generate())) }
     )
+
+    fun getWinners(): List<Car> {
+        val winnerCar = cars.maxWithOrNull(comparingInt { it.pos }) ?: throw IllegalStateException("우승자가 없을 수 없어요.")
+
+        return cars.filter { it.pos == winnerCar.pos }
```
지금 로직도 훌륭하지만 조금 더 극단적인 훈련을 해볼 수 있을 것 같아요. getter 없이 이 로직을 작성할 수 있을까요?
Car는 pos라는 상태를 갖고 있는데요, 외부에서 pos라는 상태를 모른채로 우승자를 구할 수 있다면 객체가 온전히 캡슐화되었다고 얘기할 수 있을 것 같습니다. 다음 단계에서 한 번 시도해보세요!

ddaaac reviewed (approved):
안녕하세요 기연님 👋 

이번 단계도 구현 잘 해주셔서 크게 코멘트 드릴 게 없네요.

한 가지 코멘트 남겼으니 다음 단계 진행에 참고해주세요!
